Add backend and UI for informing users about total project size in Thimble

humphd opened the pull request:
This is part of a fix for https://github.com/mozilla/thimble.mozilla.org/issues/2296.  It adds a number of new events to Bramble, and also wires up an internal system for tracking and enforcing limits on the filesystem.  I have not yet done any enforcement.

When there is activity on disk, we check to see what has happened to the total project size vs. the project's capacity, as set when the filesystem loads (default is 5MB).  We raise a `projectSizeChange` event that can be used in Thimble to display info about the current project disk usage.  When the file size goes over the capacity, we fire an event, `capacityExceeded`, and later a `capacityRestored` event if the space is recovered.  Between these events we set a flag on the `FileSystem` module that we can use in other parts of the code to block certain actions (e.g., file import, selfies, etc).  Probably that needs to happen in another patch.

flukeout commented:
Taking it for a spin now.

flukeout commented:
Okay, as far as I can tell, this is reporting things correctly and firing the "showExceeded" and "hideExceeded" events appropriately. R+ from me. Once this hits, I'll start building the indicator UI on top of it.

humphd commented:
@gideonthomas I've fixed your review issues, and refactored that code to not be one large function.  When you have time, can you re-review?

Let's land this *after* the service worker PR if possible.

humphd commented:
Reflagging @gideonthomas for review on this.

gideonthomas commented:
reviewing now

gideonthomas commented on src/bramble/client/ProjectStats.js:
```diff
@@ -1,131 +1,131 @@
                 if (!err){
-                    _cache[newPath] = _cache[oldPath];
-                    delete _cache[oldPath];
+                    self.cache.rename(oldPath, newPath);
```
what about renaming for folders?

humphd replied:
I'll file a follow-up.

humphd replied:
https://github.com/mozilla/thimble.mozilla.org/issues/2336

gideonthomas commented on src/extensions/default/bramble/lib/UI.js:
```diff
@@ -368,6 +371,35 @@ define(function (require, exports, module) {
+    function setProjectSizeInfo(info) {
+        var currentSize = Sizes.formatBytes(info.size);
+
+        // Normalize to between 0 - 1.0
```
isn't this `0 - 100.00`?

humphd commented on src/filesystem/impls/filer/lib/Sizes.js:
```diff
@@ -5,9 +5,14 @@
 define(function (require, exports, module) {
     "use strict";
 
+    var UNITS = UNITS = ['B', 'kB', 'MB'];
```
Yeah, what is that? :)

humphd commented:
I've finished this, addressed @gideonthomas' review comments, rebased on the SW stuff on master, and squashed.  I'm going to wait to land it until @flukeout has time to do the UI, since it's useless without it.

humphd commented:
Assigning to @flukeout, so we remember to come back to this and land.

flukeout commented:
Hey @humphd - I'm going to work on this now. I'll submit a PR against this branch, so let's not merge it yet.

flukeout commented:
Hey @humphd I can't get this PR to emit stuff anymore. For example I just added a basic console log...

```
    // (Re)Initialize the ProjectStats instance with a new filesystem.
    ProjectStats.prototype.init = function(fs, callback) {
        console.log("ProjectStats.init");
```

...and nothign is showing up in the console when I hit ``http://localhost:8000/src/hosted.html`` - help?

flukeout commented:
Oh wait, now I remember. I have to restart the server for changes to take effect. All good!

flukeout commented:
@humphd I'm trying to require ProjectStats into ``SidebarView.js``...

```
        ProjectStats    = require("bramble/client/ProjectStats"),
```
but when I do, the editor doesn't load and complains about 404-ing on EventEmitter.js 

I figure if I can include it in there, then I can listen for the ``projectSizeChange`` event and update the UI - does that sound like the right approach, or is there a better way to do this?

flukeout commented:
Okay, I've got a branch that has UI and a method in ``SidebarView.js`` to update what it displays. I'll ping you for help to wire it up tomorrow.

humphd commented:
You can't include `ProjectStats` inside Brackets, it lives in Thimble on the other side of the iframe.  What you want to do is put your code in the functions here: https://github.com/mozilla/brackets/pull/800/files#diff-21109d7bb135d8b29240c262a954659e, or change the functions we call here: https://github.com/mozilla/brackets/pull/800/files#diff-ff6f9f7c4c8b8efa6a41925f0fbe77d3.

humphd commented:
@flukeout some things I notice:

* We're not respecting the light-theme with this UI yet:

![screen shot 2017-07-07 at 3 02 15 pm](https://user-images.githubusercontent.com/427398/27972758-7d30ea0a-6325-11e7-87ea-ab91518afd31.png)

* Do you think it's odd that we're doing `5.24 MB`?  Probably I should change my logic on sizes to just say `5 MB`?

* Should we stop showing you your size after you exceed the max, and instead just have a string "You have run out of space, please remove some files" or something?  It's odd that I can do this:

![screen shot 2017-07-07 at 3 05 48 pm](https://user-images.githubusercontent.com/427398/27972859-cdc2a49a-6325-11e7-9b0f-3db01d8f1648.png)

* Should I add code in this PR to show you a dialog when you are over and you try to do something that would increase the size, or do it in another PR?

flukeout commented:
* I'll update to address light theme
* Yeah, 5 MB is better

humphd commented:
OK, pushed a commit to do `5 MB` vs. `5.24 MB`.

flukeout commented:
Okay, as for your other questions which I totally missed the first time:

* What happens now when your project files exceed 5 MB?
  * I think if a file you are trying to add you to exceed 5 MB, we should stop you from adding that file.
  * We can show a popup notice that says something like "You're out of room!"
* I don't think we should allow people to be in an "exceeded" state, it might be annoying or confusing if that blocks syncing / publishing etc.
  * I think in most cases it will just be the last file you added that will cause the problem. The only way out of that would be to delete that file so we may as well not even added.

We should also consider what happens when you drop in an archive that has a bunch of files that cumulatively exceed 5 MB. Should we add as many as we can until the next file would exceed 5 MB and show a popup with the files we couldn't add?

humphd commented:
I've added code to disallow any action that would cause the disk space to increase when you go over capacity; instead, I show an error dialog:

<img width="1177" alt="screen shot 2017-07-11 at 3 55 36 pm" src="https://user-images.githubusercontent.com/427398/28087741-811e7572-6651-11e7-98fd-3aee27325cd8.png">

I need some help figuring out if I've got all the actions:
* upload dialog
* drag and drop
* New File (right click in file tree)
* New file via bramble api call (e.g, `Add new HTML File` from Thimble).

@flukeout can you review again?

humphd commented:
TODO: deal with going over size when unzipping files

flukeout commented:
Reviewing now

flukeout commented:
Some thoughts so far:

* We should probably indicate to the user what actions won't be possible if a project is in an "exceeded state".
  * Can they publish it?
  * Can they update a published project if it's now exceeded?
  * Will it still save to the database?
  * They can't add new files.
  * What else?

flukeout commented:
@humphd - This is good to go, merge it!

humphd commented:
Filed https://github.com/mozilla/thimble.mozilla.org/issues/2372 and https://github.com/mozilla/thimble.mozilla.org/issues/2372 for follow-up.
